brainiak/utils/fmrisim.py

```diff
@@ -788,118 +1003,165 @@ def _calc_fwhm(volume,
+        tr = int(np.ceil(volume.shape[3] / 2))
 
-    # Calculate sfnr
-    sfnr = mean_signal / background_noise
+    # Make a matrix of brain and non_brain voxels by time
+    brain_voxels = volume[mask > 0]
+    nonbrain_voxels = volume[:, :, :, tr][mask == 0]
 
-    # Convert from memmap
-    sfnr = float(sfnr)
+    # Find the mean of the non_brain voxels (deals with structure that may
+    # exist outside of the mask)
+    nonbrain_voxels_mean = np.mean(volume[mask == 0], 1)
 
-    # What is the max activation of this volume
-    max_activity = volume.max()
+    # Take the means of each voxel over time
+    mean_voxels = np.nanmean(brain_voxels)
+    std_voxels = np.nanstd(nonbrain_voxels - nonbrain_voxels_mean)
 
-    return temporal_noise, sfnr, max_activity
+    # Return the snr
+    return mean_voxels / std_voxels
 
 
 def _calc_temporal_noise(volume,
                          mask,
                          auto_reg_order=1,
                          ):
-    """ Calculate the mix of autoregressive and drift noise.
+    """ Calculate the the temporal noise of a volume
+    This calculates the variability of the volume over time and the
+    proportion of variance over time that is due to autoregression and how
+    much is due to scanner drift.
 
     Parameters
     ----------
-    volume : 4d masarray, float
-    Input data to be calculate the autoregression
 
-    mask : 4d array, float
-    What voxels of the input are within the brain
+    volume : 4d array, float
+        Take a volume time series to extract the middle slice from the
+        middle TR
+
+    mask : 3d array, binary
+        A binary mask the same size as the volume
 
     auto_reg_order : int
-    What order of the autoregression do you want to pull out
+        What order of the autoregression do you want to pull out
+
 
     Returns
     -------
+
+
+    float 
+        The SFNR of the volume (mean brain activity divided by  temporal
+        variability in the averaged non brain voxels)  
+
     float
-    A sigma of the autoregression in the data
+        A sigma of the autoregression in the data
 
     float
-    Sigma of the drift in the data
+        Sigma of the drift in the data
 
     """
 
-    # Calculate the time course
-    timecourse = np.mean(volume[mask[:, :, :, 0] > 0], 0)
+    # Calculate sfnr and convert from memmap
+    sfnr = _calc_sfnr(volume,
+                      mask,
+                      )
+
+    # Calculate the time course of voxels within the brain
+    timecourse = np.mean(volume[mask > 0], 0)
+    demeaned_timecourse = timecourse-timecourse.mean()
 
     # Pull out the AR values (depends on order)
-    auto_reg_sigma = ar.AR_est_YW(timecourse, auto_reg_order)[1]
-    auto_reg_sigma = np.sqrt(auto_reg_sigma)
+    auto_reg_sigma = ar.AR_est_YW(demeaned_timecourse, auto_reg_order)
+    auto_reg_sigma = np.sqrt(auto_reg_sigma[1])
 
     # What is the size of the change in the time course
     drift_sigma = timecourse.std().tolist()
```

Comment:
This is an important point. You are right that this is not calculating the drift: this is not the goal. Instead I am only trying to quantify the variance attributable to risk. That is because when I later generate drift I am multiplying that time course by this value and thus describing its magnitude. There might be a better way of working with drift which I would be open to

Reply:
The FFT is a great suggestion because it will mean that I could fit for any frequency period. I could then use that value when generating drift (rather than assuming a given amount). This is a bigger change though that I will do in a later version.